pkg/volume/util/goroutinemap/goroutinemap.go

```diff
@@ -0,0 +1,252 @@
+	// and evaluate results after that.
+	Wait()
+}
+
+// NewGoroutineMap returns a new instance of GoroutineMap.
+func NewGoroutineMap(exponentialBackOffOnError bool) GoroutineMap {
+	g := &goRoutineMap{
+		operations:                []operation{},
+		exponentialBackOffOnError: exponentialBackOffOnError,
+		lock: &sync.Mutex{},
+	}
+	g.cond = sync.NewCond(g.lock)
+	return g
+}
+
+type goRoutineMap struct {
+	operations                []operation
+	exponentialBackOffOnError bool
+	cond                      *sync.Cond
+	lock                      *sync.Mutex
+}
+
+type operation struct {
+	volumeName       api.UniqueVolumeName
+	podName          types.UniquePodName
+	operationPending bool
+	expBackoff       common.ExponentialBackoff
+}
+
+func (grm *goRoutineMap) Run(
+	volumeName api.UniqueVolumeName,
+	podName types.UniquePodName,
+	operationFunc func() error) error {
+	grm.lock.Lock()
+	defer grm.lock.Unlock()
+
+	var previousOp operation
+	opExists := false
+	previousOpIndex := -1
+	for i, existingOp := range grm.operations {
+		if existingOp.volumeName != volumeName {
+			// No match, keep searching
+			continue
+		}
+
+		if existingOp.podName != emptyUniquePodName &&
+			podName != emptyUniquePodName &&
+			existingOp.podName != podName {
+			// No match, keep searching
+			continue
+		}
+
+		// Match
+		opExists = true
+		previousOp = existingOp
+		previousOpIndex = i
+		break
+	}
+
+	if opExists {
+		// Operation already exists
+		if previousOp.operationPending {
+			// Operation is pending
+			operationName := getOperationName(volumeName, podName)
+			return common.NewAlreadyExistsError(operationName)
+		}
+
+		if !previousOp.expBackoff.SafeToRetry() {
+			operationName := getOperationName(volumeName, podName)
+			return common.NewExponentialBackoffError(
+				operationName, previousOp.expBackoff)
+		}
+	}
+
+	if opExists {
+		// Update existing operation to mark as pending.
+		grm.operations[previousOpIndex].operationPending = true
+		grm.operations[previousOpIndex].volumeName = volumeName
+		grm.operations[previousOpIndex].podName = podName
+	} else {
+		// Create a new operation
+		grm.operations = append(grm.operations,
+			operation{
+				operationPending: true,
+				volumeName:       volumeName,
+				podName:          podName,
+				expBackoff:       previousOp.expBackoff,
+			})
+	}
+
+	go func() (err error) {
+		// Handle unhandled panics (very unlikely)
+		defer k8sRuntime.HandleCrash()
+		// Handle completion of and error, if any, from operationFunc()
+		defer grm.operationComplete(volumeName, podName, &err)
+		// Handle panic, if any, from operationFunc()
+		operationName := getOperationName(volumeName, podName)
+		defer common.RecoverFromPanic(operationName, &err)
+		return operationFunc()
+	}()
+
+	return nil
+}
+
+func (grm *goRoutineMap) getOperation(
+	volumeName api.UniqueVolumeName,
+	podName types.UniquePodName) (uint, error) {
+
+	for i, op := range grm.operations {
```

Comment:
@saad-ali @smarterclayton I seen now we should comment that we assume the lock is held when we do this.